ddtrace/writer.py

```diff
@@ -112,27 +114,51 @@ def _on_shutdown(self):
                     time.sleep(0.05)
 
     def _target(self):
+        result_traces = None
+        result_services = None
+
         while True:
             traces = self._trace_queue.pop()
             if traces:
                 # If we have data, let's try to send it.
                 try:
-                    self.api.send_traces(traces)
+                    result_traces = self.api.send_traces(traces)
                 except Exception as err:
                     log.error("cannot send spans: {0}".format(err))
 
             services = self._service_queue.pop()
             if services:
                 try:
-                    self.api.send_services(services)
+                    result_services = self.api.send_services(services)
                 except Exception as err:
                     log.error("cannot send services: {0}".format(err))
 
             elif self._trace_queue.closed():
-                # no traces and the queue is closed. our work is done.
+                # no traces and the queue is closed. our work is done
                 return
 
-            time.sleep(1) # replace with a blocking pop.
+            log_level = log.debug
+            if result_traces and getattr(result_traces, "status", None) >= 400:
+                now = time.time()
+                if now > self._last_error_ts + LOG_ERR_INTERVAL:
+                    log_level = log.error
+                    self._last_error_ts = now
+                log_level("failed_to_send traces to Agent: HTTP error status %s, reason %s, message %s",
+                          getattr(result_traces, "status", None), getattr(result_traces, "reason", None),
+                          getattr(result_traces, "msg", None))
+                result_traces = None
+
+            if result_services and getattr(result_services, "status", None) >= 400:
+                now = time.time()
+                if now > self._last_error_ts + LOG_ERR_INTERVAL:
+                    log_level = log.error
+                    self._last_error_ts = now
+                log_level("failed_to_send services to Agent: HTTP error status %s, reason %s, message %s",
+                          getattr(result_services, "status", None), getattr(result_services, "reason", None),
+                          getattr(result_services, "msg", None))
+                result_services = None
+
+            time.sleep(1)  # replace with a blocking pop.
```

Comment:
it's better to avoid any kind of `sleep()` in application code... why do you need that?